rust/lance-encoding/src/encodings/physical/bitpack_fastlanes.rs

```diff
@@ -0,0 +1,1547 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow::datatypes::{
+    Int16Type, Int32Type, Int64Type, Int8Type, UInt16Type, UInt32Type, UInt64Type, UInt8Type,
+};
+use arrow_array::{Array, PrimitiveArray};
+use arrow_schema::DataType;
+use bytes::Bytes;
+use futures::future::{BoxFuture, FutureExt};
+use log::trace;
+use snafu::{location, Location};
+
+use lance_arrow::DataTypeExt;
+use lance_core::{Error, Result};
+
+use crate::buffer::LanceBuffer;
+use crate::compression_algo::fastlanes::BitPacking;
+use crate::data::{DataBlock, FixedWidthDataBlock, NullableDataBlock};
+use crate::decoder::{PageScheduler, PrimitivePageDecoder};
+use crate::encoder::{ArrayEncoder, EncodedArray};
+use crate::format::ProtobufUtils;
+use arrow::array::ArrayRef;
+use bytemuck::cast_slice;
+const ELEMS_PER_CHUNK: u64 = 1024;
+
+// Compute the compressed_bit_width for a given array of integers
+// todo: compute all statistics before encoding
+// todo: see how to use rust macro to rewrite this function
```

Comment:
Here's a fun bit of Arrow devilry:

```
fn bits_needed<T: ArrowPrimitiveType>(arrays: &[ArrayRef]) -> u64
where
    T: ArrowNumericType,
    PrimitiveArray<T>: From<Vec<<T as ArrowPrimitiveType>::Native>>,
    <T as ArrowPrimitiveType>::Native:
        BitOr<Output = <T as ArrowPrimitiveType>::Native> + ArrowNativeTypeOp,
{
    let maxes = arrays
        .iter()
        .map(|arr| {
            let primitive_array = arr.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
            arrow::compute::bit_or(primitive_array).unwrap_or(T::default_value())
        })
        .collect::<Vec<_>>();
    let maxes = PrimitiveArray::<T>::from(maxes);
    let global_max = arrow::compute::bit_or(&maxes).unwrap_or(T::default_value());
    let global_max = global_max.to_usize().unwrap_or(usize::MAX) as u64;
    let num_bits =
        arrays[0].data_type().byte_width() as u64 * 8 - global_max.leading_zeros() as u64;
    // we will have constant encoding later
    if num_bits == 0 {
        1
    } else {
        num_bits
    }
}
```

Reply:
yeah, the `pub fn compute_compressed_bit_width_for_non_neg` here can definitely be simplified and write better.

but I am not sure about this code, I kind of need `global_max` to be the same type as input data type,  for example, when input is: 
```
a Datatype::Int8 array of [1, 2, 3, 4], here the `global_max` is 7(`0000 0111` in binary), when casting it to u64 and call `global_max.leading_zeros()`, the result is `61`,
which may cause the result of `num_bits` be wrong and sometimes this line
`let num_bits = arrays[0].data_type().byte_width() as u64 * 8 - global_max.leading_zeros() as u64;` may panic due to subtract overflow